tests/device_tests/test_session_id_and_passphrase.py

```diff
@@ -0,0 +1,241 @@
+    assert isinstance(response, messages.Features)
+    response = client.call_raw(
+        messages.GetPublicKey(address_n=parse_path("44'/0'/0'"), coin_name="Bitcoin")
+    )
+    assert isinstance(response, messages.PublicKey)
+    assert (
+        response.xpub
+        == "xpub6BiVtCpG9fQPxnPmHXG8PhtzQdWC2Su4qWu6XW9tpWFYhxydCLJGrWBJZ5H6qTAHdPQ7pQhtpjiYZVZARo14qHiay2fvrX996oEP42u8wZy"
+    )
+
+    # In case we want to add a new passphrase we need to send session_id = None.
+    response = client.call_raw(messages.Initialize(session_id=None))
+    assert isinstance(response, messages.Features)
+    response = client.call_raw(
+        messages.GetPublicKey(address_n=parse_path("44'/0'/0'"), coin_name="Bitcoin")
+    )
+    assert isinstance(response, messages.ButtonRequest)
+    client.debug.input("A")  # Input empty passphrase.
+    response = client.call_raw(messages.ButtonAck())
+    assert isinstance(response, messages.PublicKey)
+    assert (
+        response.xpub
+        == "xpub6CekxGcnqnJ6osfY4Rrq7W5ogFtR54KUvz4H16XzaQuukMFZCGebEpVznfq4yFcKEmYyShwj2UKjL7CazuNSuhdkofF4mHabHkLxCMVvsqG"
+    )
+
+
+@pytest.mark.skip_ui
+@pytest.mark.skip_t2
+@pytest.mark.setup_client()
+def test_passphrase_on_device_not_possible_on_t1(client):
+    # Let's start the communication by calling Initialize.
+    response = client.call_raw(messages.Initialize())
+    assert isinstance(response, messages.Features)
+
+    # Turn on passphrase.
+    _enable_passphrase(client)
+
+    # This setting makes no sense on T1.
+    response = client.call_raw(messages.ApplySettings(passphrase_always_on_device=True))
+
+    assert isinstance(response, messages.Failure)
+    assert response.code == 3  # DataError
+
+    response = client.call_raw(
+        messages.GetPublicKey(address_n=parse_path("44'/0'/0'"), coin_name="Bitcoin")
+    )
+    assert isinstance(response, messages.PassphraseRequest)
+    response = client.call_raw(messages.PassphraseAck(on_device=True))
+    assert isinstance(response, messages.Failure)
+    assert response.code == 3  # DataError
+
+
+@pytest.mark.skip_ui
+@pytest.mark.setup_client(passphrase=True)
+def test_passphrase_ack_mismatch(client):
+    response = client.call_raw(
+        messages.GetPublicKey(address_n=parse_path("44'/0'/0'"), coin_name="Bitcoin")
+    )
+    assert isinstance(response, messages.PassphraseRequest)
+    response = client.call_raw(messages.PassphraseAck(passphrase="A", on_device=True))
+    assert isinstance(response, messages.Failure)
+    assert response.code == 3  # DataError
+
+
+@pytest.mark.skip_ui
+@pytest.mark.setup_client(passphrase=True)
+def test_passphrase_missing(client):
+    response = client.call_raw(
+        messages.GetPublicKey(address_n=parse_path("44'/0'/0'"), coin_name="Bitcoin")
+    )
+    assert isinstance(response, messages.PassphraseRequest)
+    response = client.call_raw(messages.PassphraseAck(passphrase=None))
```

Comment:
maybe just `PassphraseAck()` ?

maybe also add a test case for `PassphraseAck(passphrase=None, on_device=False)` ?

Reply:
> maybe just PassphraseAck() ?

I kind of like it that it is explicit.

> maybe also add a test case for PassphraseAck(passphrase=None, on_device=False) ?

Will add.